# Chromatin_Features_Analysis
 A suite of scripts for analyzing DNA damage/repair/mutagenesis with respect to chromatin features like transcription factors and histone modifications
***

## Table of Contents
1. [Project Overview](#project-overview)
2. [Dependencies and File System Naming Conventions](#dependencies-and-file-system-naming-conventions)
3. [Preparing Data for Analysis](#preparing-data-for-analysis)
4. [Analysis and Figure Generation](#analysis-and-figure-generation)
5. [Data Availability](#data-availability)
***

## Project Overview

#### Accompanying Literature
The code in this repository was primarily used to produce the findings in [NO LINK YET](). This paper contains additional insight into the analysis using this code.

#### Background
Much of this analysis builds on [previous work](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8349767/) examining mutation rates relative to nucleosome positions. This project utilizes the software from that previous work in tandem with a suite of new scripts to examine DNA damage and repair data relative to nucleosomes in Drosophila as well as stratifying that data across [five distinct chromatin domains](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3119929/). As a result, the python scripts contained in this repository are directed at organizing genomic features such as nucleosomes with respect to these chromatin domains in preparation for comparing them against DNA damage and repair rates. The R scripts and notebook in this repository are then used to format and plot the data.
***

## Dependencies and File System Naming Conventions

#### A Disclaimer...
Admittedly, the software in this repository was largely not designed with other users in mind. Despite this, the core analysis is still very much reproducible after taking a few steps to set up a specific environment and following a few conventions pertaining to the project's file system structure. In the event that the following steps are insufficient to reproduce the desired results, you are more than welcome to [post an issue](../../issues) or email me directly at b.morledge-hampton@wsu.edu.

#### Dependencies
Besides the code in this repository, the analysis requires that the following packages are installed (The most up-to-date version of each package is recommended):
- Python:
  - [mutperiod](https://github.com/bmorledge-hampton19/mutperiod)
  - [benbiohelpers](https://github.com/bmorledge-hampton19/benbiohelpers) (Note that this should install automatically with mutperiod).
- R:
  - [data.table](https://cran.r-project.org/web/packages/data.table/index.html)
  - [ggplot2](https://cran.r-project.org/web/packages/ggplot2/index.html)
- Must be available through command line (Recommend installing through apt, where possible):
  - [bedtools](https://bedtools.readthedocs.io/en/latest/)
  - [bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
  - [samtools](http://www.htslib.org/)
  - [trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)
  
Although not required, the scripts in [this repository](https://github.com/bmorledge-hampton19/XR-seq_Analysis) may be helpful for [trimming](https://github.com/bmorledge-hampton19/XR-seq_Analysis/blob/main/TrimAdaptorSequences.py) and [aligning](https://github.com/bmorledge-hampton19/XR-seq_Analysis/blob/main/TrimmedFastqToSam.py) sequencing reads and [parsing them to bed](https://github.com/bmorledge-hampton19/XR-seq_Analysis/blob/main/SamToBed.py). If desired, these three operations can be chained together using the [AlignXRSeqReads.py](https://github.com/bmorledge-hampton19/XR-seq_Analysis/blob/main/AlignXRSeqReads.py) script.

#### Directory structure naming conventions
After cloning the chromatin_features_analysis repository, you will need to create a directory at the top level called "data". Within this data directory, two directories must be present:
- __RNAseq/__ This directory will contain RPKM values for later analysis.
- __chromatin_domains/drosophila/__  This directory will contain bed-formatted files of regions labeled as one of the five colored chromatin domains.

##### The bioinformatics directory
When running the R notebook contained in this repository, it expects data to be contained within a specific bioinformatics directory. This is merely the parent directory containing this cloned repository as well as data from mutperiod (i.e. one directory up from "chromatin_features_analysis/"). Typically, this directory is called "Bioinformatics_Projects/". Once this directory is selected, the underlying file system is inferred using the naming conventions in this section.

##### The mutperiod_data directory
Data generated by mutperiod is stored in its own directory independent of the chromatin_features_analysis project directory. This directory is created the first time mutperiod requests file input. If possible, this directory should be created within "Bioinformatics_Projects/mutperiod/". (The directory that will be created is "Bioinformatics_Projects/mutperiod/mutperiod_data/".) Next, a "drosophila_data/" directory should be created under "mutperiod_data/" and the following four directories should be created under "drosophila_data/":
- __damage/__ This directory will contain nucleosome periodicity results for cellular DNA CPD-seq data.
- __naked/__ This directory will contain nucleosome periodicity results for naked DNA CPD-seq data.
- __nucleosomes/__ This directory will contain nucleosome periodicity (repeat length) data.
- __repair/__ This directory will contain nucleosome periodicity results for XR-seq data.

Finally, under the "mutperiod_data/\_\_external_data\" directory (which should have been created automatically), create a directory named "dm6/gene_designations". This is where you will store information pertaining to the genic regionsn of the dm6 genome.

#### Data file naming conventions
When obtaining fastq sequencing files (or sam alignment files), a strict file naming system needs to be maintained in order for the rest of the pipeline to run without errors. The file names should be made up of the following identifying information, separated by underscores:
- Organism or cell type ID. Although admittedly inconsistent, this analysis expects that repair data uses "Dm" for this ID while damage uses "WT".
- Lesion type ("CPD", "6-4", or "cisplatin")
- Timepoint (in shorthand format, such as "10m" for minutes or "1h" for hours)
- Repitition number (In the format "rep#" where '#' is the repitition number or "all" when combined)

Here is an example of a valid sam file name: "Dm_CPD_1h_all.sam"

#### Deviating from naming conventions
It is still possible to run the analysis without following these instructions to the letter. This structure is imposed so that the R notebook within this repository can be run with minimal user input. However, different naming conventions can be used as long as the corresponding names within the notebook are changed.

***

## Preparing Data for Analysis

#### ???
f
***

## Analysis and Figure Generation
If the data has been prepared correctly, running the analyses and generating figures is easy: Simply run/knit the [R notebook](R/) in the repository. Also, when running the notebook, keep in mind the following details:

#### Defining the "Bioinformatics_Projects" Directory
The notebook requires a path to the "Bioinformatics_Projects" directory. (More information on this directory is available in the [bioinformatics directory](#the-bioinformatics-directory) section.) In brief, this is the parent directory to the cloned chromatin_features_analysis repository. If you are running the R notebooks in interactive mode, you will be able to select this directory through a file browser dialog. If you are knitting the notebook, you will select this directory from a list of choices in the Shiny UI for the notebook's parameters. If the desired directory is not listed, you will need to add it manually to the R notebook (on line 9).

#### Figure Output
When knitting figure generation notebooks, the figures are automatically output to the R/Drosophila_chromatin_domains_analysis_output directory, which will be created if it doesn't already exist.
***

## Data Availability
The data used for the analyses supported by this repository can be found at the following locations:
- Drosophila CPD-seq data: [NO-LINK-YET](NO-LINK-YET)
- Drosophila XR-seq data: []()
- Drosophila nucleosome map: []()
- Drosophila chromatin domain designations: []()
- Drosophila gene region designations: []()
- Drosophila RPKM values: []()
